src/frequenz/client/dispatch/recurrence.py

```diff
@@ -0,0 +1,223 @@
+
+        return pb_criteria
+
+
+# pylint: disable=too-many-instance-attributes
+@dataclass(kw_only=True)
+class RecurrenceRule:
+    """Ruleset governing when and how a dispatch should re-occur.
+
+    Attributes follow the iCalendar specification (RFC5545) for recurrence rules.
+    """
+
+    frequency: Frequency = Frequency.UNSPECIFIED
+    """The frequency specifier of this recurring dispatch."""
+
+    interval: int = 0
+    """How often this dispatch should recur, based on the frequency."""
+
+    end_criteria: EndCriteria | None = None
+    """When this dispatch should end.
+
+    Can recur a fixed number of times or until a given timestamp."""
+
+    byminutes: list[int] = field(default_factory=list)
+    """On which minute(s) of the hour the event occurs."""
+
+    byhours: list[int] = field(default_factory=list)
+    """On which hour(s) of the day the event occurs."""
+
+    byweekdays: list[Weekday] = field(default_factory=list)
+    """On which day(s) of the week the event occurs."""
+
+    bymonthdays: list[int] = field(default_factory=list)
+    """On which day(s) of the month the event occurs."""
+
+    bymonths: list[int] = field(default_factory=list)
+    """On which month(s) of the year the event occurs."""
+
+    @classmethod
+    def from_protobuf(cls, pb_rule: PBRecurrenceRule) -> "RecurrenceRule":
+        """Convert a protobuf recurrence rule to a recurrence rule.
+
+        Args:
+            pb_rule: The protobuf recurrence rule to convert.
+
+        Returns:
+            The converted recurrence rule.
+        """
+        return RecurrenceRule(
+            frequency=Frequency(pb_rule.freq),
+            interval=pb_rule.interval,
+            end_criteria=(
+                EndCriteria.from_protobuf(pb_rule.end_criteria)
+                if pb_rule.HasField("end_criteria")
+                else None
+            ),
+            byminutes=list(pb_rule.byminutes),
+            byhours=list(pb_rule.byhours),
+            byweekdays=[Weekday(day) for day in pb_rule.byweekdays],
+            bymonthdays=list(pb_rule.bymonthdays),
+            bymonths=list(pb_rule.bymonths),
+        )
+
+    def to_protobuf(self) -> PBRecurrenceRule:
+        """Convert a recurrence rule to a protobuf recurrence rule.
+
+        Returns:
+            The converted protobuf recurrence rule.
+        """
+        pb_rule = PBRecurrenceRule()
+
+        pb_rule.freq = self.frequency.value
+        pb_rule.interval = self.interval
+        if self.end_criteria is not None:
+            pb_rule.end_criteria.CopyFrom(self.end_criteria.to_protobuf())
+        pb_rule.byminutes.extend(self.byminutes)
+        pb_rule.byhours.extend(self.byhours)
+        pb_rule.byweekdays.extend([day.value for day in self.byweekdays])
+        pb_rule.bymonthdays.extend(self.bymonthdays)
+        pb_rule.bymonths.extend(self.bymonths)
+
+        return pb_rule
+
+    def prepare(self, start_time: datetime) -> rrule.rrule:
```

Comment:
This method names sounds super strange to me. If I understand correctly this method is converting the `RecurrenceRule` object to a `rrule` object. If this is correct, I would name this `as_rrule()` to be more familiar with what we use in `quantities` for example, or `to_rrule()`, to match the `to_protobuf()`. I'm also wondering if we really want to make `rrule` part of the public API. I don't remember how do we use the `rrule`, but maybe we can expose `rrule` functionality directly without exposing that we are using `rrule` internally?

Not a big deal though, I guess the client is still pretty low level so if it makes sense I'm fine with just returning the `rrule`.

Reply:
Well, this method is pretty much only used with in the public facing methods of Dispatch, we can also make it private. 
I am fine with either of your name suggestions otherwise..

Reply:
OK, I see. I guess it is the same with `from` and `to_protobuf`, they should also probably be private. What I ended up doing in the microgrid API was to just put conversion functions for protobuf in a separate (private) file, not as methods. This way the interface used by the end user doesn't need to know anything about protobuf, not even import the protobuf stuff, and also you don't need to add `# pylint: disable-...` when calling private methods. The implementation needed to do the conversion can just import the private module and that's it. For example [`microgrid_info.py`](https://github.com/frequenz-floss/frequenz-client-microgrid-python/pull/94/files#diff-3825aa4b69c70280e62cf5075d568aa48cd7fe64a4773e0434c15e0b0f5fed2c) and [`_microgrid_info_proto.py`](https://github.com/frequenz-floss/frequenz-client-microgrid-python/pull/94/files#diff-35cfa99408a5eb83d1628910b29145c1b3f12a31f9e64d17c34b5791e960bf5e).

So, not for this PR but an idea for the future. As of what to do now, I think the lowest hanging fruit would be to just rename to make it more clear and leave it public, as the other internal methods are also public currently, so if we want to fix it, all will need to be fixed.

Reply:
Once more I just noticed you moved forward with updating the code before I can reply, so probably my reply there doesn't make sense anymore :laughing: